### PR TITLE
AG-13304 - Sparkline performance improvements pt3.

### PR DESCRIPTION
**Pull request by alantreadway**
https://ag-grid.atlassian.net/browse/AG-13304

### CODE REVIEW

File: packages/ag-charts-community/src/api/preset/sparkline.ts [hunk 1]
```diff
@@ -350,7 +348,7 @@ export function sparkline(opts: AgSparklineOptions): AgCartesianChartOptions {
     };
     const yAxisBase: AgCartesianAxisOptions = {
         ...axis(yAxis, 'number'),
-        gridLine: gridLine(xAxis, true),
+        gridLine: gridLine(xAxis, false),
```
Comment (alantreadway) [inline packages/ag-charts-community/src/api/preset/sparkline.ts:351]: Disabling grid-lines means we can skip axis tick generation, so I think this makes sense from a defaulting perspective.

File: packages/ag-charts-community/src/module/optionsModule.ts
Comment (alantreadway) [inline packages/ag-charts-community/src/module/optionsModule.ts:1]: The core rework is here - I've tried to minimise the number of times we copy/clone data-structures (as this seems to be EXPENSIVE), and re-ordered the processing to facilitate this.

I've also applied micro-optimisations of removing lambdas in favor of real functions/methods, which also seems to have yielded performance improvements.

File: packages/ag-charts-community/src/util/json.ts [hunk 2]
```diff
@@ -111,26 +117,30 @@ export function shallowClone<T>(source: T): T {
...
  * @param json to traverse
  * @param visit callback for each non-primitive and non-array object found
- * @param opts
- * @param opts.skip property names to skip when walking
- * @param jsons to traverse in parallel
+ * @param skip property names to skip when walking
+ * @param parallelJson to traverse in parallel
  */
-export function jsonWalk<T>(json: T, visit: (...nodes: T[]) => void, opts?: { skip?: string[] }, ...jsons: T[]) {
+export function jsonWalk<T, C>(
```
Comment (alantreadway) [inline packages/ag-charts-community/src/util/json.ts:123]: More micro-optimisations - mostly removing use of `...` and `Array.includes()` in preference of a single param and `Set.has()` respectively (also removal of lambda use)

File: packages/ag-charts-community/src/api/preset/sparkline.ts (hunk 1 above)
Comment (jacobp100) [reply packages/ag-charts-community/src/api/preset/sparkline.ts:351]: That's fine - it's getting turned off anyway as a design requirement

File: packages/ag-charts-community/src/chart/series/cartesian/cartesianSeries.ts [hunk 3]
```diff
@@ -143,7 +143,7 @@ export interface CartesianSeriesNodeDataContext<
     visible: boolean;
 }
 
-export const RENDER_TO_OFFSCREEN_CANVAS_THRESHOLD = 10;
+export const RENDER_TO_OFFSCREEN_CANVAS_THRESHOLD = 20;
```
Comment (jacobp100) [inline packages/ag-charts-community/src/chart/series/cartesian/cartesianSeries.ts:146]: I was thinking this should be like 1000

Comment (alantreadway) [reply packages/ag-charts-community/src/chart/series/cartesian/cartesianSeries.ts:146]: I went there originally, but backed-off to make a Sparklines-specific change for now.

**Review by jacobp100 — APPROVED** [review]

File: packages/ag-charts-community/src/api/agCharts.ts [hunk 4]
```diff
@@ -263,7 +263,7 @@ class AgChartsInternal {
                     }
                 }
             },
-            { skip: ['data'] }
+            new Set('data')
```
Comment (iMoses) [inline packages/ag-charts-community/src/api/agCharts.ts]:
```suggestion
            new Set(['data'])
```
Missing array wrap, meaning the set is equal to `{ 'd', 'a', 't' }`

File: packages/ag-charts-community/src/chart/themes/chartTheme.ts [hunk 5]
```diff
@@ -396,28 +396,30 @@ export class ChartTheme {
...
             }
-        });
+        }
+    }
+
+    templateTheme<T>(themeTemplate: T, clone = true): T {
+        const themeInstance = clone ? deepClone(themeTemplate) : themeTemplate;
+        const params = this.getTemplateParameters();
+
+        jsonWalk(themeInstance, this.templateThemeFn, undefined, undefined, params);
```
Comment (iMoses) [inline packages/ag-charts-community/src/chart/themes/chartTheme.ts]:
```suggestion
        jsonWalk(themeInstance, this.templateThemeFn, undefined, params);
```
Redundant `undefined` arguments?

Comment (alantreadway) [reply packages/ag-charts-community/src/chart/themes/chartTheme.ts]: Good spot - I'm intrigued that this didn't flag up on any tests 🤔

**Review by iMoses — CHANGES_REQUESTED** [review]
One bug found, few nits.

Another nit: We now pass some private function directly to new methods, and the only reason it works without issues is that these functions don't access their context, which is broken in such cases as the methods lose their binding.
How about we make these methods static to prevent future bugs/confusion or binding of these methods, which will defeat the purpose of optimisation?

File: packages/ag-charts-community/src/util/json.ts [hunk 6]
```diff
@@ -111,26 +117,31 @@ export function shallowClone<T>(source: T): T {
...
-        });
+        visit(json, parallelJson, ctx);
+        let index = 0;
+        for (const node of json) {
+            jsonWalk(node, visit, skip, (parallelJson as any[])?.[index], ctx);
+            index++;
+        }
     } else if (isPlainObject(json)) {
-        visit(json, ...jsons);
+        visit(json, parallelJson, ctx);
```
Comment (iMoses) [inline packages/ag-charts-community/src/util/json.ts:138]: Why do we need `ctx` here?

File: packages/ag-charts-community/src/api/agCharts.ts (hunk 4 above)
Comment (alantreadway) [reply packages/ag-charts-community/src/api/agCharts.ts]: Thanks, good spot

File: packages/ag-charts-community/src/chart/themes/chartTheme.ts (hunk 5 above)
Comment (alantreadway) [reply packages/ag-charts-community/src/chart/themes/chartTheme.ts]: Actually this is correct, we want to pass the same `ctx` object into the handling function

Comment (alantreadway) [reply packages/ag-charts-community/src/chart/themes/chartTheme.ts]: Reverted

File: packages/ag-charts-community/src/util/json.ts (hunk 6 above)
Comment (alantreadway) [reply packages/ag-charts-community/src/util/json.ts:138]: This is for the `ChartTheme.templateTheme()` case where we need to pass around the map of values to apply - so we need a common 'context' that gets passed in to every `visit()` callback.

File: packages/ag-charts-community/src/chart/themes/chartTheme.ts (hunk 5 above)
Comment (iMoses) [reply packages/ag-charts-community/src/chart/themes/chartTheme.ts]: Oh, so `params` is the `ctx`...

**Review by iMoses — APPROVED** [review]
LGTM

**Pull request merged**